[TEST] remove createExpectedQuery from BaseQueryTestCase

javanna opened the pull request:
We currently test the toQuery method by comparing its return value with the output of createExpectedQuery. This seemed at first like a good deep test for the lucene generated queries, but ends up causing a lot of code duplication between tests and prod code, which smells like this test is not that useful after all.

This commit removes the requirement for createExpectedQuery, we test a bit less the lucene generated queries, but we do still have a testToQuery method in BaseQueryTestCase. This test acts as smoke test to verify that there are no issues with our toQuery method implementation for each query: it verifies that two equal queries should result in the same lucene query. It also verifies that changing the boost value of a query affects the result of the toQuery.

Part of the previous createExpectedQuery can be moved to assertLuceneQuery using normal assertions, we do it when it makes sense only though.

This commit also adds support for boost and queryName to SpanMultiTermQueryParser and SpanMultiTermQueryBuilder, plus it removes no-op setters for queryName and boost from QueryFilterBuilder. Instead we simply declare in our test infra that query filter doesn't support boost and queryName and we disable tests aroudn these fields in that specific case. Boost and queryName support is also moved down to the QueryBuilder interface.

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -219,49 +221,72 @@ public void testFromXContent() throws IOException {
     /**
-     * Creates the expected lucene query given the current {@link QueryBuilder} and {@link QueryParseContext}.
-     * The returned query will be compared with the result of {@link QueryBuilder#toQuery(QueryParseContext)} to test its behaviour.
+     * Few queries do allow you to se the boost and queryName but don't do anything with it. This method allows
```
typo s/se/set

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -283,7 +308,9 @@ public void testSerialization() throws IOException {
     protected static QueryParseContext createContext() {
-        return new QueryParseContext(index, queryParserService);
+        QueryParseContext queryParseContext = new QueryParseContext(index, queryParserService);
+        queryParseContext.parseFieldMatcher(ParseFieldMatcher.EMPTY);
```
Just curious about what this does.

javanna replied:
it prevents an NPE because we don't set the ParseFieldMatcher in our test context otherwise.

cbuescher replied:
thanks

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/MissingQueryBuilderTest.java:
```diff
@@ -68,96 +63,8 @@ private String getRandomFieldName() {
-        final boolean existence = queryBuilder.existence();
-        final boolean nullValue = queryBuilder.nullValue();
-        String fieldPattern = queryBuilder.fieldPattern();
+    protected void doAssertLuceneQuery(MissingQueryBuilder queryBuilder, Query query, QueryParseContext context) throws IOException {
```
Not sure if given the complexity of the query creation here it is possible to do something similar as in ExistsQueryBuilderTest here for example?

javanna replied:
right, I think you can clearly ee that I gave up on this query... I can look again but it becomes too complex to test anything here... I also want to try to stay away from mappings etc.

cbuescher replied:
We can leave it then, maybe add some note to revisit it later once again to see if we can extend test a bit here?

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/RandomQueryBuilder.java:
```diff
@@ -53,6 +53,14 @@ public static QueryBuilder createQuery(Random r) {
         }
     }
 
+    public static QueryBuilder createDifferentQuery(Random r, Class<? extends QueryBuilder> queryBuilderClass) {
```
Interesting, but can't find place where this is used in this PR?

javanna replied:
leftover, will remove

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/RangeQueryBuilderTest.java:
```diff
@@ -93,32 +89,16 @@ protected RangeQueryBuilder doCreateTestQueryBuilder() {
+        if (getCurrentTypes().length == 0 || (queryBuilder.fieldName().equals(DATE_FIELD_NAME) == false && queryBuilder.fieldName().equals(INT_FIELD_NAME) == false)) {
+            assertThat(query, instanceOf(TermRangeQuery.class));
+        } else if (queryBuilder.fieldName().equals(DATE_FIELD_NAME)) {
+            //we can't properly test unmapped dates because LateParsingQuery is package private
```
Should we increase visibility in this case? Makes only sense if we can then really test more, though, otherwise okay to leave it I guess.

javanna replied:
LateParsingQuery is a PITA in general... making it public would make things worse I think?

cbuescher replied:
Then we should leave it, the situation in general improves here already.

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/SimpleQueryStringBuilderTest.java:
```diff
@@ -177,15 +178,6 @@ public void testNullQueryTextGeneratesException() {
     }
 
-    @Test
-    public void testHandlingDefaults() throws IOException {
```
Was this test not useful anymore? Maybe just test the builder here.

javanna replied:
to be honest I don't get what this test was doing... hence I didn't know what to replace it with given that createExpectedQuery is gone...

cbuescher replied:
I think what it was supposed to test is that `toQuery()` uses reasonable defaults if some parameters (like analyzer) are not set in the builder. However, it seems that this is already covers in the randomized builder setup, so if we just want to check that `toQuery` produces something in these cases, this should be covered.

cbuescher commented:
Left a few comments, in general prefer the idea here of not testing the generated queries to deeply in our unit tests here since this should happen (and already happens) in intergration tests elsewhere. However not that we are back to the beginning where we wrote individual assertions for the generation lucene queries in `assertLuceneQuery` we will again have to decide in each individual test which level of detail we want to test there. Maybe some general guideline here would help (maybe in the docs), e.g. the method should at least test generated query class type and if possible the effect some builder parameters have on the resulting query, or nested query collection size etc... This should just serve to shorten discussion when working on susequent query tests.

javanna commented:
thanks for the review @cbuescher I addressed your comments. As for the need for guidelines...I think what we test will depend on the query, it can be discussed case by case as part of the review process. We do want to try and test as much as possible, but we don't want to end up writing useless tests as we did up until now :) the trade-off really depends on the query.

cbuescher commented:
Went through the changes again, LGTM now.
